odl/solvers/advanced/chambolle_pock.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""First-order primal-dual algorithm developed by Chambolle and Pock.
+
+The method is proposed in [CP2011a]_ and augmented with diagonal
+preconditioners in [CP2011b]_. The algorithm is flexible and well apt suited
+for non-smooth, convex optimization problems in imaging. This implementation
+is along the lines of [SJP2012]_.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+# External
+import numpy as np
+
+# Internal
+from odl.operator.operator import Operator
+from odl.solvers.util import Partial
+
+__all__ = ('chambolle_pock_solver',)
+
+
+# TODO: add dual gap as convergence measure
+# TODO: diagonal preconditioning
+# TODO: preferred way to hyperlink References, see Reference section
+
+def chambolle_pock_solver(op, x, tau, sigma, proximal_primal, proximal_dual,
+                          theta=1, gamma=None, niter=1, partial=None,
+                          x_relax=None, y=None):
```

Comment:
Too many parameters. Try to decide which are advanced and put them in `kwargs`. You can also make a section `Other Parameters` in the docstring, which will appear after the examples (I think) in the HTML doc.